File: superstitioMain/java/superstitio/characters/Maso.java
package superstitio.characters;

import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.badlogic.gdx.math.MathUtils;
import com.megacrit.cardcrawl.cards.AbstractCard;
import com.megacrit.cardcrawl.characters.AbstractPlayer;
import com.megacrit.cardcrawl.core.CardCrawlGame;
import com.megacrit.cardcrawl.dungeons.Exordium;
import com.megacrit.cardcrawl.screens.CharSelectInfo;
import com.megacrit.cardcrawl.screens.charSelect.CharacterOption;
import superstitio.DataManager;
import superstitio.Logger;
import superstitio.cards.CardOwnerPlayerManager;
import superstitio.cards.general.BaseCard.Kiss;
import superstitio.cards.general.BaseCard.Masturbate;
import superstitio.cards.maso.BaseCard.FistIn;
import superstitio.cards.maso.BaseCard.Invite_Maso;
import superstitio.relics.a_starter.DoubleBlockWithVulnerable;
import superstitio.relics.blight.DevaBody_Masochism;
import superstitio.relics.blight.JokeDescription;
import superstitio.relics.blight.MasochismMode;
import superstitioapi.player.PlayerInitPostDungeonInitialize;
import superstitioapi.renderManager.characterSelectScreenRender.RenderInCharacterSelect;

import java.util.ArrayList;

import static com.megacrit.cardcrawl.dungeons.AbstractDungeon.*;
import static superstitio.DataManager.SPTT_DATA.MasoEnums.MASO_CARD;
import static superstitio.DataManager.SPTT_DATA.MasoEnums.MASO_Character;
import static superstitioapi.relicToBlight.InfoBlight.addAsInfoBlight;

// 继承CustomPlayer类
public class Maso extends BaseCharacter implements PlayerInitPostDungeonInitialize, RenderInCharacterSelect {
    public static final String ID = DataManager.MakeTextID(Maso.class.getSimpleName());
    public static final CharacterSelectInfo characterInfo = new CharacterSelectInfo(60, 70, 99);

    public Maso(String name) {
        super(ID, name, MASO_Character);
    }

    public static void setUpMaso() {
        if (floorNum > 1 || !(CardCrawlGame.dungeon instanceof Exordium)) return;
        player.currentHealth = player.getLoadout().currentHp;
        if (ascensionLevel >= 6) {
            player.currentHealth = MathUtils.round((float) player.currentHealth * 0.9F);
        }
    }

    protected static ArrayList<String> MasoStartDeck() {
        ArrayList<String> startingDeck = new ArrayList<>();
        for (int x = 0; x < 5; x++) {
            startingDeck.add(Kiss.ID);
        }
        for (int x = 0; x < 4; x++) {
            startingDeck.add(Invite_Maso.ID);
        }
        startingDeck.add(Masturbate.ID);
        startingDeck.add(FistIn.ID);
        return startingDeck;
    }

    @Override
    public int getAscensionMaxHPLoss() {
        return 5;
    }

    @Override
    // 初始遗物
    public ArrayList<String> getStartingRelics() {
        ArrayList<String> retVal = new ArrayList<>();
        retVal.add(DoubleBlockWithVulnerable.ID);
        return retVal;
    }

    @Override
    public CharSelectInfo getLoadout() {
        return new CharSelectInfo(characterStrings.NAMES[0], // 人物名字
                characterStrings.TEXT[0], // 人物介绍
                characterInfo.currentHp, // 当前血量
                characterInfo.maxHp, // 最大血量
                0, // 初始充能球栏位
                110, // 初始携带金币
                5, // 每回合抽牌数量
                this, // 别动
                this.getStartingRelics(), // 初始遗物
                this.getStartingDeck(), // 初始卡组
                false // 别动
        );
    }

    @Override
    public ArrayList<String> getStartingDeck() {
        Logger.run("Begin loading starter Deck Strings");
        return MasoStartDeck();
    }

    // 你的卡牌颜色（这个枚举在最下方创建）
    @Override
    public AbstractCard.CardColor getCardColor() {
        return MASO_CARD;
    }

    // 创建人物实例，照抄
    @Override
    public AbstractPlayer newInstance() {
        return new Maso(this.name);
    }

    @Override
    public void initPostDungeonInitialize() {
//        if (!SuperstitioConfig.isEnableGuroCharacter()) {
//            SuperstitioConfig.setEnableGuroCharacter(true);
//        }
        setUpMaso();
        addAsInfoBlight(new JokeDescription());
        addAsInfoBlight(new DevaBody_Masochism());
        addAsInfoBlight(new MasochismMode());
    }

    @Override
    public void renderInCharacterSelectScreen(CharacterOption characterOption, SpriteBatch sb) {

    }

    @Override
    public void updateInCharacterSelectScreen(CharacterOption characterOption) {
        unableByGuroSetting();
    }

    @Override
    protected boolean isCardCanAdd(AbstractCard card) {
        return CardOwnerPlayerManager.isMasoCard(card);
    }
}

File: superstitioMain/java/superstitio/characters/Maso.kt
package superstitio.characters

import com.badlogic.gdx.graphics.g2d.SpriteBatch
import com.badlogic.gdx.math.MathUtils
import com.megacrit.cardcrawl.cards.AbstractCard
import com.megacrit.cardcrawl.cards.AbstractCard.CardColor
import com.megacrit.cardcrawl.characters.AbstractPlayer
import com.megacrit.cardcrawl.core.CardCrawlGame
import com.megacrit.cardcrawl.core.EnergyManager
import com.megacrit.cardcrawl.dungeons.AbstractDungeon
import com.megacrit.cardcrawl.dungeons.Exordium
import com.megacrit.cardcrawl.screens.CharSelectInfo
import com.megacrit.cardcrawl.screens.charSelect.CharacterOption
import superstitio.DataManager
import superstitio.DataManager.SPTT_DATA.MasoEnums
import superstitio.Logger
import superstitio.cards.CardOwnerPlayerManager
import superstitio.cards.general.BaseCard.Kiss
import superstitio.cards.maso.BaseCard.FistIn
import superstitio.cards.maso.BaseCard.Invite_Maso
import superstitio.cards.maso.BaseCard.Spark
import superstitio.relics.a_starter.VulnerableTogetherRelic
import superstitio.relics.blight.DevaBody_Masochism
import superstitio.relics.blight.EnjoyAilment
import superstitio.relics.blight.JokeDescription
import superstitio.relics.blight.MasochismMode
import superstitioapi.player.PlayerInitPostDungeonInitialize
import superstitioapi.relicToBlight.InfoBlight
import superstitioapi.renderManager.characterSelectScreenRender.RenderInCharacterSelect

// 继承CustomPlayer类
class Maso(name: String) : BaseCharacter(ID, name, MasoEnums.MASO_Character), PlayerInitPostDungeonInitialize,
    RenderInCharacterSelect
{
    init
    {
        // 初始化你的人物，如果你的人物只有一张图，那么第一个参数填写你人物图片的路径。
        this.initializeClass(
            MASO_CHARACTER,  // 人物图片
            MASO_CHARACTER_SHOULDER_2, MASO_CHARACTER_SHOULDER_2, BlondHairBlueEyes_CORPSE_IMAGE,  // 人物死亡图像
            this.loadout, 0.0f, 0.0f, 250.0f, 375.0f,  // 人物碰撞箱大小，越大的人物模型这个越大
            EnergyManager(3) // 初始每回合的能量
        )
    }

    override fun getAscensionMaxHPLoss(): Int
    {
        return 5
    }

    // 初始遗物
    override fun getStartingRelics(): ArrayList<String>
    {
        val retVal = ArrayList<String>()
        retVal.add(VulnerableTogetherRelic.ID)
        return retVal
    }

    override fun getLoadout(): CharSelectInfo
    {
        return CharSelectInfo(
            characterStrings!!.NAMES[0],  // 人物名字
            characterStrings.TEXT[0],  // 人物介绍
            characterInfo.currentHp,  // 当前血量
            characterInfo.maxHp,  // 最大血量
            0,  // 初始充能球栏位
            characterInfo.gold,  // 初始携带金币
            5,  // 每回合抽牌数量
            this,  // 别动
            this.startingRelics,  // 初始遗物
            this.startingDeck,  // 初始卡组
            false // 别动
        )
    }

    override fun getStartingDeck(): ArrayList<String>
    {
        Logger.run("Begin loading starter Deck Strings")
        return MasoStartDeck()
    }

    // 你的卡牌颜色（这个枚举在最下方创建）
    override fun getCardColor(): CardColor
    {
        return MasoEnums.MASO_CARD
    }

    // 创建人物实例，照抄
    override fun newInstance(): AbstractPlayer
    {
        return Maso(this.name)
    }

    override fun initPostDungeonInitialize()
    {
        setUpMaso()
        InfoBlight.addAsInfoBlight(JokeDescription())
        InfoBlight.addAsInfoBlight(DevaBody_Masochism())
        InfoBlight.addAsInfoBlight(MasochismMode())
        InfoBlight.addAsInfoBlight(EnjoyAilment())
    }

    override fun renderInCharacterSelectScreen(characterOption: CharacterOption, sb: SpriteBatch)
    {
    }

    override fun updateInCharacterSelectScreen(characterOption: CharacterOption)
    {
        unableByGuroSetting()
    }

    override fun isCardCanAdd(card: AbstractCard?): Boolean
    {
        return CardOwnerPlayerManager.isMasoCard(card)
    }

    companion object
    {
        val ID: String = DataManager.MakeTextID(Maso::class.java.simpleName)
        val characterInfo: CharacterSelectInfo = CharacterSelectInfo(60, 70, 110)

        fun setUpMaso()
        {
            if (AbstractDungeon.floorNum > 1 || CardCrawlGame.dungeon !is Exordium) return
            AbstractDungeon.player.currentHealth = AbstractDungeon.player.loadout.currentHp
            if (AbstractDungeon.ascensionLevel >= 6)
            {
                AbstractDungeon.player.currentHealth =
                    MathUtils.round(AbstractDungeon.player.currentHealth.toFloat() * 0.9f)
            }
        }

        fun MasoStartDeck(): ArrayList<String>
        {
            val startingDeck = ArrayList<String>()
            for (x in 0..4)
            {
                startingDeck.add(Kiss.ID)
            }
            for (x in 0..3)
            {
                startingDeck.add(Invite_Maso.ID)
            }
            //        startingDeck.add(Masturbate.ID);
            startingDeck.add(Spark.ID)
            startingDeck.add(FistIn.ID)
            return startingDeck
        }
    }
}